openhtf/core/test_state.py

```diff
@@ -130,7 +146,53 @@ def test_api(self):
                 running_phase_state.attachments,
                 running_phase_state.attach,
                 running_phase_state.attach_from_file,
-                self.notify_update))
+                self.get_measurement,
+                self.get_attachment,
+                self.notify_update,))
+
+  def get_measurement(self, measurement_name):
+    """Get a copy of a measurement value from current or previous phase.
+
+    Args:
+      measurement_name: str of the measurement name
+    Returns:
+      an ImmutableMeasurement
+    """
+    # The framework ignores measurements from SKIP and REPEAT phases
+    ignore_outcomes = set([test_record.PhaseOutcome.SKIP,])
+
+    # Check current running phase state
+    if self.running_phase_state:
+      if measurement_name in self.running_phase_state.measurements:
+        return ImmutableMeasurement.FromMeasurement(
+            self.running_phase_state.measurements[measurement_name])
+
+    # Iterate through phases in reversed order for LIFO.
+    for phase_record in reversed(self.test_record.phases):
+      if (phase_record.result not in ignore_outcomes and
+          measurement_name in phase_record.measurements):
+        measurement = phase_record.measurements[measurement_name]
+        return ImmutableMeasurement.FromMeasurement(measurement)
+
+    test.logger.warning('Could not find measurement: %s', measurement_name)
+    return None
+
+  def get_attachment(self, attachment_name):
+    """Get a copy of an attachment from current or previous phase."""
+    for phase_record in self.test_record.phases:
+      if attachment_name in phase_record.attachments:
+        attachment = phase_record.attachments[attachment_name]
+        return copy.deepcopy(attachment)
```

Comment:
attachments are by-design potentially quite large, i'm not sure the implicit deepcopy in what is otherwise seemingly a simple getter method is particularly user-friendly.  if there isn't a concrete need for get_attachment() i'd remove it.

Reply:
So had a long discussion with Farz about this in this PR conversation.  Technically you can attach a mutable data structure (like a dict) and I'm trying to prevent users from mutating the attachment.

i.e. a test author shouldn't do something like this:

```
# in PhaseOne
test.attach('my_data', {data: [1, 2, 3] })

# in PhaseTwo
extra_data = [4, 5, 6]
attachment.get_attachment('my_data')
attachment.contents.update({'extra_data': extra_data})

```